src/EditorFeatures/Core/Shared/Preview/PreviewSolutionCrawlerRegistrationService.cs

```diff
@@ -68,6 +74,34 @@ public void Unregister(Workspace workspace, bool blockingShutdown = false)
             {
                 source.Cancel();
```

Comment:
This appears to be a race condition.  We're asking to cooperatively cancel.  But we then go and clear out data even though the other threads may still be computing it.  How do we know that we won't end up calling "RemoveDocument" but then have diagnostics cached that hold onto the Workspace?

It seems like we want to cancel the work, then wait for the existing work to finish before proceeding. 

One way to do this would be to make it so that Register(Worksoace) calls a helper method which is Task returning.  You could then have a dictionary that maps from Workspace to that task.  Unregister would then cancel the token, then grab the task and wait until it completes.  It would then go and clear all the cached data.